File: 131-palindrome-partitioning/palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def isPalindrome(left,right):
            while left <= right:
                if s[left] != s[right]:
                    return False
                left+=1
                right-=1
            return True
        def fun(i,arr):
            if i == len(s):
                result.append(arr.copy())
                return
            for ctr in range(i+1,len(s)+1):
                if isPalindrome(i,ctr-1):
                    arr.append(s[i:ctr])
                    fun(ctr,arr)
                    arr.pop(-1)
        result = []
        fun(0,[])
        return result

File: 131-palindrome-partitioning/palindrome-partitioning.java
/**

aab

a      aa


 */

class Solution {
    List<List<String>> result;
    String S;
    public List<List<String>> partition(String s) {
        List<String> curr = new ArrayList<>();
        S = s;
        result = new ArrayList<>();
        recFun(0,0,new ArrayList<>());
        return result;
    }
    public void recFun(int start, int end, List<String> curr){
        if(start == S.length() && end == S.length()){
            result.add(new ArrayList<>(curr));
            return;
        }
        if(end == S.length()){
            return;
        }
        // if start to end (inclusive) is a palindrome we recursively call from end + 1 to last
        if(isPalindrome(start,end)){
            curr.add(S.substring(start,end+1));
            recFun(end + 1,end+1,curr);
            curr.remove(curr.size()-1);
        }
        recFun(start,end+1,curr);

    }
    public boolean isPalindrome(int start, int end){
        while(start <= end){
            if(S.charAt(start) != S.charAt(end)){
                return false;
            }
            start+=1;
            end-=1;
        }
        return true;
    }
}